File: dotnet/examples/Pactman.NonprofitCheckPlus.Examples/Cases/Ex20BulkBatchLimits.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Threading.Tasks;
using Pactman.NonprofitCheckPlus.Dev;
using Pactman.NonprofitCheckPlus.Exceptions;

namespace Pactman.NonprofitCheckPlus.Examples.Cases;

/// <summary>
/// EX-20 — Bulk batch limits.
/// </summary>
/// <remarks>
/// The 50-EIN ceiling, why the SDK will not chunk for you, and how to chunk deliberately.
/// </remarks>
public sealed class Ex20BulkBatchLimits : IExample
{
    /// <inheritdoc />
    public string Id => "ex-20";

    /// <inheritdoc />
    public string Title => "The batch ceiling, and chunking a larger list deliberately.";

    /// <inheritdoc />
    public async Task RunAsync(string[] args)
    {
        using var context = ExampleContext.WithFixtures();

        Output.Heading("The limit");
        Output.Field("Endpoints.MaxBulkEins", Endpoints.MaxBulkEins);

        Output.Heading("Over the limit");

        var oversized = Enumerable.Range(0, Endpoints.MaxBulkEins + 1)
            .Select(index => index.ToString("000000000", CultureInfo.InvariantCulture))
            .ToArray();

        try
        {
            await context.Client.Nonprofits.CheckBulkAsync(oversized);

            throw new ExampleFailedException("An oversized batch was accepted. It should have been rejected.");
        }
        catch (PactmanValidationException error)
        {
            // Local, before the request: an oversized batch costs nothing.
            Output.Field("type", error.GetType().Name);
            Output.Field("origin", error.Origin.ToWireValue());
            Output.Field("message", error.Message);
        }

        Output.Heading("Chunking, in caller code");

        // The SDK does not chunk automatically, because chunking decides how much quota
        // a call spends. That is the caller's decision to make and to see in the diff.
        var portfolio = Fixtures.KnownEins.ToList();
        const int batchSize = 5;

        var batches = Chunk(portfolio, batchSize).ToList();

        Output.Field("organizations", portfolio.Count);
        Output.Field("batch size", batchSize);
        Output.Field("requests", batches.Count);

        var matched = 0;
        var missed = new List<string>();

        foreach (var batch in batches)
        {
            var result = await context.Client.Nonprofits.CheckBulkAsync(batch);

            matched += result.Organizations.Count;
            missed.AddRange(result.NotFoundEins);
        }

        Output.Field("matched", matched);
        Output.Field("not found", missed.Count);

        Output.Note(
            "Chunk deliberately, and size the batch against your rate limit rather than\n"
            + "against the maximum. Fifty at a time is allowed; it is not always wise.");
    }

    private static IEnumerable<IReadOnlyList<string>> Chunk(IReadOnlyList<string> items, int size)
    {
        for (var offset = 0; offset < items.Count; offset += size)
        {
            yield return items.Skip(offset).Take(size).ToList();
        }
    }
}

File: java/examples/src/main/java/org/pactman/nonprofitcheckplus/examples/cases/Ex20BulkBatchLimits.java
package org.pactman.nonprofitcheckplus.examples.cases;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import org.pactman.nonprofitcheckplus.Endpoints;
import org.pactman.nonprofitcheckplus.devtools.FixtureEins;
import org.pactman.nonprofitcheckplus.examples.support.Example;
import org.pactman.nonprofitcheckplus.examples.support.ExampleContext;
import org.pactman.nonprofitcheckplus.examples.support.ExampleFailedException;
import org.pactman.nonprofitcheckplus.examples.support.Output;
import org.pactman.nonprofitcheckplus.exceptions.PactmanValidationException;
import org.pactman.nonprofitcheckplus.models.BulkCheckResult;

/**
 * EX-20 — The batch limit, and chunking a larger list yourself.
 *
 * <p>The SDK will not split an oversized batch for you. Splitting turns one call
 * a caller intended into several billable calls they did not, so it is a
 * decision the caller makes.
 */
public final class Ex20BulkBatchLimits implements Example {

    @Override
    public String id() {
        return "ex-20";
    }

    @Override
    public String title() {
        return "The 50-EIN batch limit, and chunking on your own terms";
    }

    @Override
    public void run(String[] args) {
        try (ExampleContext context = ExampleContext.withFixtures()) {
            Output.heading("The limit is enforced locally");
            Output.field("Endpoints.MAX_BULK_EINS", Endpoints.MAX_BULK_EINS);

            List<String> tooMany =
                    new ArrayList<>(Collections.nCopies(
                            Endpoints.MAX_BULK_EINS + 1, FixtureEins.PUBLIC_CHARITY));

            try {
                context.client().nonprofits().checkBulk(tooMany);
                throw new ExampleFailedException("An oversized batch should have been rejected.");
            } catch (PactmanValidationException failure) {
                Output.field("sent", 0);
                Output.field("message", failure.getMessage());
            }

            Output.heading("Chunking a larger list");

            // A real workload would hold hundreds of EINs. Chunking is four
            // lines, and doing it here means the caller can see how many
            // billable requests they are about to make.
            List<String> everything = new ArrayList<>();

            for (int i = 0; i < 6; i++) {
                everything.addAll(java.util.Arrays.asList(
                        FixtureEins.PUBLIC_CHARITY,
                        FixtureEins.PUBLIC_CHARITY_SECOND,
                        FixtureEins.PRIVATE_FOUNDATION));
            }

            int batchSize = 5;
            int batches = 0;
            int matched = 0;

            for (int start = 0; start < everything.size(); start += batchSize) {
                List<String> chunk =
                        everything.subList(start, Math.min(start + batchSize, everything.size()));
                BulkCheckResult result = context.client().nonprofits().checkBulk(chunk);
                batches += 1;
                matched += result.getOrganizations().size();
            }

            Output.field("EINs", everything.size());
            Output.field("batch size", batchSize);
            Output.field("requests made", batches);
            Output.field("organizations returned", matched);
        }

        Output.note("The batch size above is 5 to keep the example short; the API's limit is 50\n"
                + "and a real workload should use it. Note that duplicates within a chunk are\n"
                + "still sent — pass dedupe(true), or clean the input, if the list came from\n"
                + "somewhere you do not control.");
    }
}
